hcl/ast/merged_block.go

```diff
@@ -39,31 +43,57 @@ type MergedBlock struct {
 	RawBlocks map[string]Blocks
 }
 
+// BlockType represents a block type.
+type BlockType string
+
+// LabelBlockType represents a labelled block type.
+type LabelBlockType struct {
+	Type   BlockType // Type of the block
+	Labels string    // Labels are comma separated.
+}
+
 // MergedBlocks maps the block name to the MergedBlock.
 type MergedBlocks map[string]*MergedBlock
 
+// MergedLabelBlocks maps the block labels/types to the MergedBlock.
+type MergedLabelBlocks map[LabelBlockType]*MergedBlock
+
 // NewMergedBlock creates a new MergedBlock of type typ.
-func NewMergedBlock(typ string) *MergedBlock {
+func NewMergedBlock(typ string, labels []string) *MergedBlock {
 	return &MergedBlock{
-		Type:       typ,
+		Type:       BlockType(typ),
+		Labels:     strings.Join(labels, "."),
 		Attributes: make(Attributes),
 		Blocks:     make(map[string]*MergedBlock),
 		RawBlocks:  make(map[string]Blocks),
 	}
 }
 
+// NewLabelBlockType returns a new LabelBlockType.
+func NewLabelBlockType(typ string, labels []string) LabelBlockType {
+	return LabelBlockType{
+		Type:   BlockType(typ),
+		Labels: strings.Join(labels, "."),
```

Comment:
maybe we could extract a newLabels that does the joining, the "." as the separator is duplicated here.